imap/imapparse.c

```diff
@@ -259,160 +259,112 @@ EXPORTED int getxstring(struct protstream *pin, struct protstream *pout,
     return EOF;
 }
 
-EXPORTED int getint32(struct protstream *pin, int32_t *num)
+/* can't flag with -1 if there is no number here, so return EOF */
+static int _getint(struct protstream *pin, uint64_t max, uint64_t *num)
 {
-    int32_t result = 0;
+    uint64_t result = 0;
+    uint64_t tens = max / 10;
+    uint8_t ones = max % 10;
     int c = EOF;
     int gotchar = 0;
 
-    /* INT_MAX == 2147483647 */
     while ((c = prot_getc(pin)) != EOF && cyrus_isdigit(c)) {
-        if (result > 214748364 || (result == 214748364 && (c > '7')))
+        uint8_t digit = c - '0';
+
+        if (result > tens || (result == tens && (digit > ones)))
```

Comment:
Why not just test if it's over INT32_MAX and ignore the tens/ones logic entirely?  We have TONS of headroom with the 64 bits!

Reply:
This is just a straight port of what was there. And there is no headroom for getuint64()

Reply:
Rewritten to use strtoull() as we were doing in getmodseq()